tensorflow/core/ops/image_ops.cc

```diff
@@ -131,6 +131,71 @@ Status NMSShapeFn(InferenceContext* c) {
   return Status::OK();
 }
 
+Status CombinedNMSShapeFn(InferenceContext* c) {
+  //Get inputs and validate ranks
+  ShapeHandle boxes;
+  //boxes is a tensor of Dimensions [batch_size, num_anchors, q, 4]
+  TF_RETURN_IF_ERROR(c->WithRank(c->input(0), 4, &boxes));
+  ShapeHandle scores;
+  //scores is a tensor of Dimensions [batch_size, num_anchors, num_classes]
+  TF_RETURN_IF_ERROR(c->WithRank(c->input(1), 3, &scores));
+  ShapeHandle max_output_size_per_class;
+  TF_RETURN_IF_ERROR(c->WithRank(c->input(2), 0, &max_output_size_per_class));
+  ShapeHandle max_total_size;
+  TF_RETURN_IF_ERROR(c->WithRank(c->input(3), 0, &max_total_size));
+  ShapeHandle unused_shape;
+  TF_RETURN_IF_ERROR(c->WithRank(c->input(4), 0, &unused_shape));
+  TF_RETURN_IF_ERROR(c->WithRank(c->input(5), 0, &unused_shape));
+
+  DimensionHandle unused;
+  //boxes[0] and scores[0] are both batch_size
+  TF_RETURN_IF_ERROR(c->Merge(c->Dim(boxes, 0), c->Dim(scores, 0), &unused));
+  //boxes[1] and scores[1] are both num_anchors
+  TF_RETURN_IF_ERROR(c->Merge(c->Dim(boxes, 1), c->Dim(scores, 1), &unused));
+  // The boxes[3] is 4.
+  TF_RETURN_IF_ERROR(c->WithValue(c->Dim(boxes, 3), 4, &unused));
+
+  DimensionHandle d = c->Dim(boxes, 2);
+  auto num_classes = c->Value(c->Dim(scores, 2));
+  if (c->ValueKnown(d)) {
+    if (c->Value(d) != 1 && c->Value(d) != num_classes) {
+       return errors::InvalidArgument("third dimension of boxes must be either "
+           "1 or equal to the third dimension of scores");
+    }
+  }
+  DimensionHandle output_dim;
+  DimensionHandle batch_dim = c->Dim(boxes, 0);
+
+  TF_RETURN_IF_ERROR(c->MakeDimForScalarInput(3, &output_dim));
+  if (c->ValueKnown(output_dim) && c->Value(output_dim) <= 0) {
+    return errors::InvalidArgument("max_total_size should be > 0 ");
```

Comment:
should we just default to size_per_class * num_classes. a negative value would mean a user does not intended to cap.
we can check that par_per_class is true.

Reply:
https://github.com/tensorflow/models/blob/master/research/object_detection/builders/post_processing_builder.py#L78
Fails if max_total_size is less than size_per_class. Even if pad_per_class is true, we still need max_total_size as it might be smaller than size_per_class * num_classes. 
If the user does not intend to cap, they can set pad_per_class to true and max_total_size to > size_per_class * num_classes. Does that sound reasonable?